File: src/Class/ObjectBase.c
#include "ObjectBase.r"

/**
 * @brief Destructs `ObjectBase`.
 *
 * @param objectBase The `ObjectBase` to destruct.
 * @see Destructable
 */
void ObjectBaseDestructor(ObjectBase *objectBase) {
    free(objectBase);
}

void constructor_ObjectBase_fields(ObjectBase *objectBase) {
    static Constructable const constructable = {
            .constructor = (void *(*const)(void) )(&ObjectBaseConstructor)};
    objectBase->constructable = &constructable;

    static Destructable const destructable = {
            .destructor = (void *(*const)(void *) )(&ObjectBaseDestructor)};
    objectBase->destructable = &destructable;
}

ObjectBase *ObjectBaseConstructor() {
    ObjectBase *obj = malloc(sizeof *obj);
    if (obj == NULL) { /* error handling here */
    }

    obj->CLASS_NAME         = "ObjectBase";

    constructor_ObjectBase_fields(obj);

    return obj;
}

File: src/Class/ObjectBase.r
#ifndef CLASS_OBJECTBASE_R
#define CLASS_OBJECTBASE_R

#include "Constructable.r"
#include "Destructable.r"
#include <stdlib.h>

// Forward declaration of incomplete type
typedef struct objectBase ObjectBase;

struct objectBase {
    const char *CLASS_NAME;

    const Constructable *constructable;

    const Destructable *destructable;
};

ObjectBase *ObjectBaseConstructor();

#endif //CLASS_OBJECTBASE_R
